File: day12/cxx/puzzle2.cxx
#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include "range.hpp"
#include "util.hpp"

using namespace std;

const auto solve()
{
    ifstream f{"../input.txt"};

    auto terrain = load_terrain(f);
    auto map = std::get<0>(terrain);
    auto end = std::get<2>(terrain);

    AStar a_star {map};

    auto shortest = map.size() * map[0].size();

    for (auto y: iter_range(0, map.size()))
    {
        for (auto x: iter_range(0, map[y].size()))
        {
            if (map[y][x] == 0)
            {
                auto path_len = a_star.solve({x, y}, end);

                if (path_len > 0 && path_len < shortest)
                {
                    shortest = path_len;
                }

            }
        }
    }

    return shortest;
}

int main()
{
    auto solution = solve();
    cout << solution << endl;
}

File: day12/py/puzzle2.py
#!/usr/bin/env python
from typing import TextIO, Tuple, List

from util import load_terrain, AStar
            
def solve():

    with open('../input.txt', 'r') as f:
        map, start, end = load_terrain(f)
        a_star = AStar(map)

    shortest = len(map) * len(map[0])

    for y in range(0, len(map)):
        for x in range(0, len(map[y])):
            if map[y][x] == 0:
                path_len = a_star.solve((x, y), end)

                if path_len is not None and path_len < shortest:
                    shortest = path_len

    return shortest

if __name__ == "__main__":
    solution = solve()
    print(solution)

